File: gd_payment_sdk/android/src/main/kotlin/net/geidea/gd_payment_sdk/GdPaymentResultHandler.kt
// © 2026 Geidea. Proprietary and confidential.
// Unauthorized copying or redistribution is prohibited.

package net.geidea.gd_payment_sdk

import io.flutter.plugin.common.MethodChannel.Result
import net.geidea.sdk.sdk.GDPaymentError
import net.geidea.sdk.sdk.GDPaymentResult

class GdPaymentResultHandler {

    fun handleSuccess(paymentResult: GDPaymentResult, result: Result) {
        val resultMap = buildSuccessMap(paymentResult)
        result.success(resultMap)
    }

    fun handleFailure(error: GDPaymentError, result: Result) {
        val details = buildErrorDetails(error)
        result.error(error.code, error.message, details)
    }

    fun handleCancellation(result: Result) {
        val resultMap = mapOf("status" to "canceled")
        result.success(resultMap)
    }

    private fun buildSuccessMap(paymentResult: GDPaymentResult): Map<String, Any?> {
        return mapOf(
            "status" to "success",
            "data" to mapOf(
                "orderId" to paymentResult.orderId,
                "tokenId" to paymentResult.tokenId,
                "agreementId" to paymentResult.agreementId,
                "paymentMethod" to paymentResult.paymentMethod?.let { buildPaymentMethodMap(it) }
            )
        )
    }

    private fun buildPaymentMethodMap(paymentMethod: Any): Map<String, Any?> {
        // Cast to the actual PaymentMethodResult type
        val pm = paymentMethod as? net.geidea.sdk.sdk.PaymentMethodResult ?: return emptyMap()

        return mapOf(
            "type" to pm.type,
            "brand" to pm.brand,
            "cardholderName" to pm.cardholderName,
            "maskedCardNumber" to pm.maskedCardNumber,
            "wallet" to pm.wallet,
            "expiryDate" to pm.expiryDate?.let { buildExpiryDateMap(it) }
        )
    }

    private fun buildExpiryDateMap(expiryDate: Any): Map<String, Any?> {
        val exp = expiryDate as? net.geidea.sdk.sdk.ExpiryDateResult ?: return emptyMap()

        return mapOf(
            "month" to exp.month,
            "year" to exp.year
        )
    }

    private fun buildErrorDetails(error: GDPaymentError): Map<String, Any?> {
        return mapOf(
            "status" to "failure",
            "code" to error.code,
            "message" to error.message,
            "details" to error.details
        )
    }
}

File: gd_payment_sdk/ios/Classes/GdPaymentResultHandler.swift
// © 2026 Geidea. Proprietary and confidential.
// Unauthorized copying or redistribution is prohibited.

import Flutter
import GeideaPaymentSDK

public class GdPaymentResultHandler {

    func handleSuccess(result: GeideaPaymentSDK.GDPaymentResult, flutterResult: @escaping FlutterResult) {
        let resultMap = buildSuccessMap(from: result)
        flutterResult(resultMap)
    }

    func handleFailure(error: GeideaPaymentSDK.GDSDKError, flutterResult: @escaping FlutterResult) {
        let details = buildErrorDetails(from: error)

        flutterResult(FlutterError(
            code: error.code,
            message: error.message,
            details: details
        ))
    }

    func handleCancellation(flutterResult: @escaping FlutterResult) {
        let resultMap: [String: Any] = ["status": "canceled"]
        flutterResult(resultMap)
    }

    // MARK: - Private Helper Methods

    private func buildSuccessMap(from result: GeideaPaymentSDK.GDPaymentResult) -> [String: Any] {
        var dataMap: [String: Any] = [
            "orderId": result.orderId as Any,
            "tokenId": result.tokenId as Any,
            "agreementId": result.agreementId as Any
        ]

        if let paymentMethod = result.paymentMethod {
            dataMap["paymentMethod"] = buildPaymentMethodMap(from: paymentMethod)
        }

        return [
            "status": "success",
            "data": dataMap
        ]
    }

    private func buildPaymentMethodMap(from paymentMethod: Any) -> [String: Any] {
        // Cast to the actual PayPaymentMethod type from GeideaPaymentSDK
        guard let pm = paymentMethod as? GeideaPaymentSDK.PayPaymentMethod else {
            return [:]
        }

        var paymentMethodMap: [String: Any] = [:]

        // Access properties directly (not via KVO)
        if let type = pm.type {
            paymentMethodMap["type"] = type
        }

        if let brand = pm.brand {
            paymentMethodMap["brand"] = brand
        }

        if let cardholderName = pm.cardholderName {
            paymentMethodMap["cardholderName"] = cardholderName
        }

        if let maskedCardNumber = pm.maskedCardNumber {
            paymentMethodMap["maskedCardNumber"] = maskedCardNumber
        }

        // Handle wallet if it exists
        // paymentMethodMap["wallet"] = pm.wallet // Add this if wallet property exists

        if let expiryDate = pm.expiryDate {
            paymentMethodMap["expiryDate"] = buildExpiryDateMap(from: expiryDate)
        }

        return paymentMethodMap
    }

    private func buildExpiryDateMap(from expiryDate: GeideaPaymentSDK.PayPaymentMethod.ExpiryDate) -> [String: Any] {
        return [
            "month": expiryDate.month,
            "year": expiryDate.year
        ]
    }

    private func buildErrorDetails(from error: GeideaPaymentSDK.GDSDKError) -> [String: Any] {
        return [
            "status": "failure",
            "code": error.code,
            "message": error.message,
            "details": error.details as Any
        ]
    }
}
